src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.ProjectState.cs

```diff
@@ -308,21 +308,21 @@ private async Task<DiagnosticAnalysisResult> LoadInitialAnalysisDataAsync(Projec
                 {
                     cancellationToken.ThrowIfCancellationRequested();
 
-                    if (!await TryDeserializeDocumentAsync(serializer, document, builder, cancellationToken).ConfigureAwait(false))
+                    if (!await TryDeserializeDocumentAsync(persistentService, serializer, document, builder, cancellationToken).ConfigureAwait(false))
                     {
                         continue;
                     }
                 }
 
-                if (!await TryDeserializeAsync(serializer, project, project.Id, _owner.NonLocalStateName, s_addOthers, builder, cancellationToken).ConfigureAwait(false))
+                if (!await TryDeserializeAsync(persistentService, serializer, project, document: null, project.Id, _owner.NonLocalStateName, s_addOthers, builder, cancellationToken).ConfigureAwait(false))
```

Comment:
Instead of a single public method with 3 parameters, namely `Project`, `Document` and `ProjectOrDocumentId`, can we two public overloads for `TryDeserializeAsync`, one which takes `Project` and another which takes `Document` and both these invoke a private method with the 3 parameters you have currently?

Reply:
Seems `TryDeserializeAsync` is already a private method, you can ignore the suggestion.

Reply:
Yes. I think we can simplify this nevertheless. I don't think we need to pass delegates to TryDeserializeAsync.